### ISSUE TITLE
Lexical ambiguities on marking terminals should generate helpful message.

### ISSUE THREAD

**Issue by ericvanwyk** [melt-umn/copper#2]
The error message for a lexical ambiguity on marking terminals is displayed just like any other kind of lexical ambiguity.  But this is the one error message that a programmer might see when composing language extensions.  This should be a special message with some instructions on what it is, and how to fix it.  Perhaps containing a URL to good instructions and and explanation.

**Comment by schwerdf** [comment]
I need a clearer definition of what is meant by "lexical ambiguity on marking terminals." Copper imposes no constraints on how marking terminals are used in disambiguation functions, so it would be possible, during an end-user composition, to see ambiguities involving a mix of regular terminals and marking terminals.

Also, would some of this (especially the embedding of specific documentation URLs in error messages) be better implemented on the level closest to the specific use case (*i.e.*, in Silver)? In the past, we have discussed automating marking-terminal disambiguation on the Silver side, and the instructions would change, in a Silver-specific way, if this was implemented. From the standpoint of implementation, the `LexicalAmbiguityMessage` class should have enough information for Silver to detect whether a lexical ambiguity involves marking terminals or not.

**Comment by ericvanwyk** [comment]
If only extensions that pass the modular determinism analysis are used,
then the only lexical ambiguities would be between marking terminals.  So,
when this happens it would be nice to see an error message that is directed
at the end-user programmer and not at a language developer.

I'm not sure how to have Silver do this since this message doesn't show up
until Copper runs as part of the Ant script.  Isn't Silver done at this
point?

For some form of automated marking terminal disambiguation we'd need to
change how Silver and Copper interact.  Would Silver try to call Copper
directly instead of generating an ant script to do so?

Ted, do you remember a conversation about this?

Cheers,
Eric

**Comment by schwerdf** [comment]
Oh. I thought Silver had switched to using the API; never mind.

The MDA guarantees no lexical ambiguities except those *involving* marking terminals; it does not guarantee only ambiguities *between* marking terminals. The details are in section 6.6 of my thesis (which also covers the strategy of automation we had discussed).

In practice, there is also the issue of disambiguation functions, which I alluded to above. For example, see the attached patch [[1]](https://github.com/melt-umn/copper/files/212214/MDAPassing-with-disambig-function.patch.txt) to the test grammar `MDAPassing.xml`: the extension in the patched grammar passes the MDA, but its marking terminal is in a disambiguation function, so if another marking terminal with the same lexeme was introduced, host terminals would show up in the resulting ambiguity.
